wordpress-mobile/WordPress-iOS
# [Modal Layout Picker] Create sections and thumbnail UI using static data

Fixes https://github.com/wordpress-mobile/gutenberg-mobile/issues/2436

## To test:
<details>
<summary>To disable or enable the development version of Modal Layout Picker</summary>

- Open the app from the build that allows FeatureFlags such as a PR build or a local development build
    - By default, the modal layout picker will be enabled.
    - From the site page:
        - Click on your Gravatar.
        - Click on App Settings.
        - Click on Debug.
        - Toggle "Gutenberg Modal Layout Picker" to enable or disable the picker
            <details>
            <summary>To toggle:</summary>
            <img width=300 src="https://user-images.githubusercontent.com/3384451/87160515-106f5680-c291-11ea-9d49-fac5e6b7e66c.gif" />
            </details>
</details>

When enabled, the Layout Picker should show when creating a new page from My Site and from the Page list.

Start by navigating to the Modal Layout Picker

### General Navigation
_Note:_ Currently, the categories repeat the same layout 30 times because the static data only has one layout per category. 

1. Category rows should be horizontally scrollable within their sections with a preview image
    - Pick a row and scroll horizontally. 
    - Expect the category to scroll to the end and back again
1. Scrolling down and back up on the page should retain the scroll position of the category row
    - Pick a row and scroll horizontally.
    - Scroll vertically.
    - Scroll back to your reference row and expect it to be in the same location
1. Scrolling down and back up on the page should retain the selected layout
    - Select a layout
    - Scroll vertically and/or horizontally.
    - Scroll back to your reference layout and expect it to be selected still

### Preview Images
1. Layouts that haven't loaded yet should display an activity indicator indicating it's trying to load
    - On first load or when scrolling down for the first time expect to see an activity indicator on images that haven't loaded yet.
1. Images should be cached for use offline
    - Open the Layout picker and allow some images to download
    - Leave the layout picker
    - Either go offline or monitor the network calls
    - Open the layout picker
    - Expect the images for the layout picker to be pulled from the cache (no new network calls made for already loaded images).
    -  _Note:_ Because we're reusing the same image you might see multiple network calls to load the image on the first display. This is expected but you shouldn't see any network requests on subsequent loads. The static data used for the layouts can be found in the Gutenberg project in [GutenbergLayouts.swift](https://github.com/WordPress/gutenberg/blob/master/packages/react-native-bridge/ios/GutenbergLayouts.swift#L112)

### Selecting a Layout
1. Selecting a layout should highlight the cell and add a checkmark
    - Open the layout picker and select a cell

    | iOS 12 and Below | iOS 13+ | Dark Mode |
    | --- | --- | --- |
    | <img width="300" src="https://user-images.githubusercontent.com/3384451/89220323-5fae5b80-d59f-11ea-97dd-fbff3205143c.png"> | <img width="300" src="https://user-images.githubusercontent.com/3384451/89220221-2fff5380-d59f-11ea-979c-ddf8b09f77bc.png"> | <img width="300" src="https://user-images.githubusercontent.com/3384451/89220214-2b3a9f80-d59f-11ea-8823-7eeae732df9a.png"> | 

1. Only one layout should be selected at any given time
    - Open the layout picker and select a cell
    - Select a different cell (either in the same row or a different row)
    - Expect the old cell to deSelect and the new one to be selected
1. Selecting a Layout should show the buttons "Preview" and "Create Page"
    - Open the layout picker and select a cell
    - _Note:_ The "Preview" and "Create Page" buttons are non-functional right now and will be handled in a later task
1. Deselecting a layout should revert the bottom bar to "Create Blank Page"
    - Open the layout picker and select a cell
    - Deselect the cell by tapping on it again
    - Expect the bottom buttons to revert to "Create Blank Page"

## Screenshots:

| iOS 12 and Below | iOS 13+ | Dark Mode |
| --- | --- | --- |
| <img width="300" src="https://user-images.githubusercontent.com/3384451/89220323-5fae5b80-d59f-11ea-97dd-fbff3205143c.png"> | <img width="300" src="https://user-images.githubusercontent.com/3384451/89220221-2fff5380-d59f-11ea-979c-ddf8b09f77bc.png"> | <img width="300" src="https://user-images.githubusercontent.com/3384451/89220214-2b3a9f80-d59f-11ea-8823-7eeae732df9a.png"> | 
| <img width="300" src="https://user-images.githubusercontent.com/3384451/89220430-8b314600-d59f-11ea-9bd5-d9f037a92a14.png"> | <img width="300" src="https://user-images.githubusercontent.com/3384451/89220460-9a17f880-d59f-11ea-94c5-c0161b1532d9.png"> | <img width="300" src="https://user-images.githubusercontent.com/3384451/89220489-a734e780-d59f-11ea-8770-10c149549b3c.png"> |

## Additional Context: 

This is a continuation of the work started in https://github.com/wordpress-mobile/WordPress-iOS/pull/14501 and will be followed up by other tickets as part of [Modal Layout Picker Project](https://github.com/wordpress-mobile/gutenberg-mobile/issues?q=is%3Aissue+is%3Aopen+%5BModal+Layout+Picker%5D)

Placeholder images and downloading the images on poor network connections will be refined as part of https://github.com/wordpress-mobile/gutenberg-mobile/issues/2445

## PR submission checklist:
